p2p/protocol/adv/adv.go

```diff
@@ -0,0 +1,122 @@
+package adv
+
+import (
+	"context"
+	"encoding/json"
+	"net"
+	"net/http"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-libp2p-core/host"
+	peer "github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	gostream "github.com/libp2p/go-libp2p-gostream"
+)
+
+var log = logging.Logger("go-legs-adv")
+
+func deriveProtocolID(topic string) protocol.ID {
+	return protocol.ID("/legs/adv/" + topic + "/0.0.1")
+}
+
+func (p *AdvPublisher) Serve(ctx context.Context, host host.Host, topic string) error {
+	l, err := gostream.Listen(host, deriveProtocolID(topic))
+	if err != nil {
+		return err
+	}
+
+	p.server = &http.Server{Handler: http.Handler(p)}
+	errChan := make(chan error)
+	go func() {
+		errChan <- p.server.Serve(l)
+	}()
+
+	select {
+	case err := <-errChan:
+		return err
+	case <-ctx.Done():
+		return p.server.Close()
+	}
+}
+
+func newHttpClient(ctx context.Context, host host.Host, topic string, peer peer.ID) (http.Client, error) {
+	transport := http.Transport{
+		DialContext: func(ctx context.Context, network, addr string) (net.Conn, error) {
+			return gostream.Dial(ctx, host, peer, deriveProtocolID(topic))
+		},
+	}
+
+	return http.Client{
+		Transport: &transport,
+	}, nil
+}
+
+type AdvClient struct {
+	client *http.Client
+}
+
+func NewAdvClient(ctx context.Context, host host.Host, topic string, peer peer.ID) (*AdvClient, error) {
+	client, err := newHttpClient(ctx, host, topic, peer)
+	return &AdvClient{
+		client: &client,
+	}, err
+}
+
+func (c *AdvClient) QueryRootCid(ctx context.Context) (cid.Cid, error) {
```

Comment:
Since we aren't keeping the http client as a statefull object, can we collapse this all to just 1 function?

You can create an http.Client with the custom transport and save some extra indirection.